### PR TITLE
Unit tests for one shot object detector

### LINKED ISSUE #1788: Unit tests for One Shot Object Detection
Assign to Project 5.6 Sprint 2 when board becomes available.

### PR DESCRIPTION
**Pull request by shreyajain17**
fixes #1788

### CODE REVIEW

**Comment by TobyRoseman** [comment]
Are we abandoning #1999?

**Comment by TobyRoseman** [comment]
How long does this test take to run?

**Comment by shreyajain17** [comment]
@TobyRoseman  12 passed in 61.26 seconds

File: src/unity/python/turicreate/test/test_one_shot_object_detector.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+                pil_img.save(f, format=image_format)
+                return tc.Image(f.name)
+
+    num_examples = 100
+    num_starter_images = 5
+    max_num_boxes_per_image = 10
+    classes = _CLASSES
+    images = []
+    FORMATS = ['png', 'jpeg', 'raw']
+    for i in range(num_examples):
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_one_shot_object_detector.py]: nit: in Python when a variable is never intended to actually be used it's standard to name the variable `_`.

File: src/unity/python/turicreate/test/test_one_shot_object_detector.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        # Model
+        self.train, self.test, self.backgrounds = _get_data(feature=self.feature, target=self.target)
+        self.model = tc.one_shot_object_detector.create(self.train,
+                                                        target=self.target,
+                                                        backgrounds=self.backgrounds,
+                                                        batch_size=2,
+                                                        max_iterations=1)
+
+        ## Answers
+        self.opts = self.def_opts.copy()
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_one_shot_object_detector.py]: It looks like `self.opts` is never getting used. . . . which means you don't need this line or the next line or the declaration of `self.def_opts`.

File: src/unity/python/turicreate/test/test_one_shot_object_detector.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        self.opts['max_iterations'] = 1
+
+        self.get_ans = {
+           'target': lambda x: x == self.target,
+           'num_starter_images': lambda x: x > 0,
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_one_shot_object_detector.py]: Why not `lambda x: x == len(self.train)`?

File: src/unity/python/turicreate/test/test_one_shot_object_detector.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            tc.one_shot_object_detector.create(self.train, target='wrong_feature', 
+                backgrounds=self.backgrounds, max_iterations=1)
+
+    def test_create_with_empty_dataset(self):
+        with self.assertRaises(_ToolkitError):
+            tc.one_shot_object_detector.create(self.train[:0], target=self.target)
+
+    def test_predict(self):
+        sf = self.test.head()
+        # Make sure this does not need the annotations column to work
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_one_shot_object_detector.py]: I don't understand this comment. The object detector's prediction function doesn't even take an annotation column.

File: src/unity/python/turicreate/test/test_one_shot_object_detector.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        img_fixed = tc.image_analysis.resize(img, 416, 416, 3)
+        pil_img = Image.fromarray(img_fixed.pixel_data)
+        if _mac_ver() >= (10, 13):
+            ret = coreml_model.predict({self.feature: pil_img}, 
+                usesCPUOnly = True)
+            self.assertEqual(ret['coordinates'].shape[1], 4)
+            self.assertEqual(ret['confidence'].shape[1], len(_CLASSES))
+            self.assertEqual(ret['coordinates'].shape[0], 
+                ret['confidence'].shape[0])
+            # A numeric comparison of the resulting of top bounding boxes is
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_one_shot_object_detector.py]: I would be very curious to understand why.

File: src/unity/python/turicreate/test/test_one_shot_object_detector.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        self.assertEqual(type(str(model)), str)
+        self.assertEqual(type(model.__repr__()), str)
+
+    def test_save_and_load(self):
+        with test_util.TempDirectory() as filename:
+            self.model.save(filename)
+            self.model = tc.load_model(filename)
+
+            self.test_predict()
+            print("Predict passed")
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_one_shot_object_detector.py]: I would remove these print statements

File: src/unity/python/turicreate/toolkits/one_shot_object_detector/one_shot_object_detector.py [hunk 7]
```diff
@@ -59,6 +59,7 @@ def create(data,
         # Make predictions on the training set and as column to the SFrame
         >>> test_data['predictions'] = model.predict(test_data)
     """
+    _tkutl._raise_error_if_column_exists(data, target, "dataset", target)
```
Comment (nickjong) [inline src/unity/python/turicreate/toolkits/one_shot_object_detector/one_shot_object_detector.py]: The variable name is `data` so the third argument, `dataset_variable_name`, should be `"data"`

File: src/unity/python/turicreate/test/test_one_shot_object_detector.py (hunk 4 above)
Comment (shreyajain17) [reply src/unity/python/turicreate/test/test_one_shot_object_detector.py]: Yeah I deleted the code but forgot to delete the comment.

File: src/unity/python/turicreate/test/test_one_shot_object_detector.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    @classmethod
+    def setUpClass(self):
+        """
+        The setup class method for the basic test case with all default values.
+        """
+        self.feature = 'myimage'
+        self.target = 'mytarget'
+
+        ## Create the model
+        self.def_opts = {
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_one_shot_object_detector.py]: This isn't being used.

**Comment by nickjong** [comment]
> I think something might be messed up with the history here; I don't think you want the changes outside of `test_one_shot_object_detector.py`

Note that the changes in `one_shot_object_detector.py` and `one_shot_object_detector.cpp` fix actual bugs revealed by the new tests.

**Comment by nickjong** [comment]
It looks like `test_export_coreml` is failing in Travis (with `std::bad_alloc`) or simply not returning (after ten minutes). Does this test succeed locally?

**Comment by znation** [comment]
From an in-person discussion with @shreyajain17, merging this now so we can get the toolkit fixes in, and we'll fix unit tests in master if they are still failing there. It's likely bad_alloc in Travis is a result of Travis being RAM-constrained, and not a real bug (we already disable other tests on Travis for this reason).

**Comment by shreyajain17** [comment]
> It looks like `test_export_coreml` is failing in Travis (with `std::bad_alloc`) or simply not returning (after ten minutes). Does this test succeed locally (on Mac and Linux)?

Yes locally passes everything on Mac

**Pull request merged**